nnc/Store.swift

```diff
@@ -147,71 +200,114 @@ private let fpzipDecode:
+      }
+      guard encodedSize[0] > 4 else { return 0 }
+      var zippedDataSize = encodedSize[0] - 4
+      guard zip(data: exponents,
+                dataSize: floatCount,
+                zippedData: encoded.advanced(by: 4),
+                zippedDataSize: &zippedDataSize) else { return 0 }
+      encoded.assumingMemoryBound(to: Int32.self)[0] = Int32(zippedDataSize)
+      guard 4 + zippedDataSize + floatCount <= encodedSize[0] else { return 0 }
+      memcpy(encoded.advanced(by: 4 + zippedDataSize), floatsWithoutExp, floatCount)
+      identifier?[0] = 0x511
+      encodedSize[0] =
+        4 /* for compressed exponents size */
+        + zippedDataSize /* exponents */
+        + floatCount /* floats without exponent */
       return 1
     }
 
-  private let zipDecode:
-    @convention(c) (
-      UnsafeRawPointer?, Int, Int32, UnsafePointer<Int32>?, Int32, UInt32, UnsafeMutableRawPointer?,
-      UnsafeMutableRawPointer?, UnsafeMutablePointer<Int>?
-    ) -> Int32 = {
-      data, dataSize, dataType, dimensions, dimensionCount, identifier, context, decoded,
-      decodedSize
-      in
-      guard identifier == 0x217 else { return 0 }
-      guard let data = data, let dimensions = dimensions, let decoded = decoded,
-        let decodedSize = decodedSize, dimensionCount > 0
-      else { return 0 }
-      let nextIn = data.assumingMemoryBound(to: UInt8.self)
-      let nextOut = decoded.assumingMemoryBound(to: UInt8.self)
-      var stream = compression_stream(
-        dst_ptr: nextOut, dst_size: decodedSize[0], src_ptr: nextIn, src_size: dataSize, state: nil)
-      var status = compression_stream_init(&stream, COMPRESSION_STREAM_DECODE, COMPRESSION_ZLIB)
-      guard status != COMPRESSION_STATUS_ERROR else { return 0 }
-      defer { compression_stream_destroy(&stream) }
-      stream.src_ptr = nextIn
-      stream.src_size = dataSize
-      stream.dst_ptr = nextOut
-      stream.dst_size = decodedSize[0]
-      repeat {
-        status = compression_stream_process(&stream, Int32(COMPRESSION_STREAM_FINALIZE.rawValue))
-        guard status != COMPRESSION_STATUS_ERROR else { return 0 }
-      } while status == COMPRESSION_STATUS_OK && stream.dst_size > 0
-      decodedSize[0] = decodedSize[0] - stream.dst_size
-      return 1
+private let ezm8Decode:
+  @convention(c) (
+    UnsafeRawPointer?, Int, Int32, UnsafePointer<Int32>?, Int32, UInt32, UnsafeMutableRawPointer?,
+    UnsafeMutableRawPointer?, UnsafeMutablePointer<Int>?
+  ) -> Int32 = {
+    data, dataSize, dataType, dimensions, dimensionCount, identifier, context, decoded, decodedSize
+    in
+    guard dataType == Int32(CCV_16F) else { return 0 }
+    guard identifier == 0x511 else { return 0 }
+    guard let data = data, let dimensions = dimensions, let decoded = decoded,
+      let decodedSize = decodedSize, dimensionCount > 0
+    else { return 0 }
+    var floatCount = decodedSize[0] / MemoryLayout<Float16>.size
```

Comment:
here i'm relying on decodedSize to be correct

Reply:
It is not per previous comment. But it should be straightly less than or equal to the original tensor size.